ENH: Add interactive quality plots to `summarize` visualization

jakereps opened the pull request:
~~**Work in progress**~~

gregcaporaso commented:
In this [paired-end quality plot](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdl.dropboxusercontent.com%2Fu%2F2868868%2Fpaired-end-demux-qual-plot.qzv), the forward and reverse read data look identical. Are you plotting the same data twice by accident?

gregcaporaso commented:
There is currently a lot of whitespace in the plots because the y-axis goes to ~100. The highest score should be 40 (as long as the user has provided the correct PHRED offset). What if we made 40 the max y value in these plots, and if we detect values out of range (we could check that the means for each position are between 40 and 0) we can display the plot with out of range values cut off (so it's abundantly clear that something is wrong), and display a warning to the user in the viz. The warning would be something like: *Some of the PHRED quality values are out of range. This is likely because the an incorrect PHRED offset was chosen on import of your raw data. You can learn how to choose your PHRED offset during import in the importing tutorial.* (The importing tutorial will discuss this in 2017.3.)

jakereps commented:
It looks like there was one line of debugging code left making it plot the forward reads twice. Resolved it now! Thanks @gregcaporaso!

gregcaporaso commented:
@benjjneb and I were both confused about how to zoom back out after zooming in. Can you include instructions on that (in the visualization, I think)? I had to ask @thermokarst, and @benjjneb had to ask me. 

@benjjneb also says that he thinks it's fine to drop the `dada2 plot-qualities` in favor of these plots, so we should do that for `2017.3`.

gregcaporaso commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -0,0 +1,144 @@
+TEMPLATES = pkg_resources.resource_filename('q2_demux', '_summarize')
+
+
+def _decode_qual_to_phred(qual_str):
```
for clarity, can you call this `_decode_qual_to_phred33`?

gregcaporaso commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -0,0 +1,144 @@
+    rev = manifest[manifest.direction == 'reverse'].filename.tolist()
+
+    per_sample_fastq_counts = {}
+    for file in fwd:
```
It will be possible for there to only be reverse reads in a single-end read file, so I think this needs to be a little smarter. You can do something along the lines of:

```python
if len(fwd) > 0:
    reads = fwd
else:
    reads = rev

per_sample_fastq_counts = {}
for file in reads:
```

gregcaporaso commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -0,0 +1,144 @@
+
+    quality_scores = collections.defaultdict(list)
+    subsample_ns = sorted(random.sample(range(1, result.sum() + 1), n))
+    target = subsample_ns.pop(0)
```
Could this be:

```python
for subsample_n in subsample_ns:
```

That should get rid of some of the duplicated logic in the for loop below where you're moving to the next target.

jakereps replied:
Oops, yeah I meant to go back at this and do a refactor, this was the "get it working" sampling method. Since we have the counts of each sample, it can actually skip entire files based on the counts if need be.

gregcaporaso commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -0,0 +1,144 @@
+    target = subsample_ns.pop(0)
+    current = 1
+    if paired:
+        for f, r in zip(sorted(fwd), sorted(rev)):
```
I'd break this out into its own function, for clarity. Same goes for the code in the `else` block.

gregcaporaso commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -0,0 +1,144 @@
+            if target == -1:
+                break
+    else:
+        for f in fwd:
```
Same comment as above about how single-end read data may only be reverse, not forward.

gregcaporaso commented:
> It looks like there was one line of debugging code left making it plot the forward reads twice. Resolved it now!

Confirmed that that fixes it, thanks @jakereps!

ebolyen commented on q2_demux/_summarize/assets/src/box.js:
```diff
@@ -0,0 +1,313 @@
+// Heavily modified variant of http://bl.ocks.org/jensgrubert/7789216
+// which is based off of https://gist.github.com/mbostock/4061502
+// which is written and marked as being under GPL-3
```
This will be an issue, as all derivatives (including this PR) would need to be GPL-3

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+  const xAxis = d3.axisBottom(x).ticks(12);
+  const yAxis = d3.axisLeft(y).ticks((12 * props.height) / props.width);
+
+  const iqr = (d) => {
```
`iqr` isn't the best name here. This is really giving you the index ranges of where the outliers would be right?

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+    .text(d => d);
+
+  const maxX = d3.max(data, d => d[0]);
+  const x0 = [0, maxX];
```
I think there's an off-by-one error either where the box-plots are placed, or in the ticks placement. (It's easy to see if you zoom in enough to cause the ticks to be floats.)

ebolyen replied:
older qzv, that looks perfect

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+    let i = -1;
+    let j = d.length;
+    while (d[++i] < q1 - iq);  // eslint-disable-line no-plusplus
+    while (d[--j] > q3 + iq);  // eslint-disable-line no-plusplus
```
This is so darn pretty :unicorn:

jakereps replied:
RIP `iqr` ⚰️

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+    .domain([0, 100]);
+
+  const brush = d3.brush();
+  let idleTimeout;
```
Is it possible to encapsulate this state/zooming/brushing into a sub-component?

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+  .enter()
+    .append('g')
+    .attr('class', 'boxplot')
+    .attr('transform', d => `translate(${x(d[0])}, ${props.margin.top})`)
```
Can you use your axis here to calculate the x-offset?

jakereps replied:
Can you clarify this, @ebolyen? It's currently using the axis in `x(d[0])`.

ebolyen replied:
I mean instead of using SVG's translate operation, since you have the axis on hand, it should be able to calculate the exact offset (if I'm remembering how this works right).

jakereps replied:
Oh, I see. `g`'s are only positionable using the `transform/translate` capabilities. `x` and `y` don't affect them. Using the `g`'s for each boxplot allows relative placement for the svg elements themselves, so they can be placed without even knowing their actual location, so the `x1`, `x2`, etc... in the plots can just be 0 instead of needing the axis to calculate anything.

ebolyen replied:
That makes perfect sense :+1:

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+    .append('g')
+    .attr('class', 'boxplot')
+    .attr('transform', d => `translate(${x(d[0])}, ${props.margin.top})`)
+    .call(chart.width((x.range()[1] - x.range()[0]) / (x.domain()[1] - x.domain()[0]) / 2));
```
Can this `width` call be hoisted up out of `call`s expression? If I understand what is happening, I think you can get away with calling the factory method once and then reusing the resulting function.

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+      .attr('dy', '1em')
+      .attr('font-size', '10px')
+      .style('text-anchor', 'middle')
+      .text('Sequence Base');
```
This is just stylistic, so you can ignore, but I usually see the axis setup at the top of a plotting function in D3.

jakereps replied:
I have some magic white rectangles boxing in the plot, otherwise on zooming it shows beyond the axis, so this will have to be added after those blocks in order to be visible, and those blocks will have to be after the boxes in order to block them, since `svg` is entirely order based making this stuck as-is.

ebolyen commented on q2_demux/_summarize/assets/src/plot.js:
```diff
@@ -0,0 +1,229 @@
+    .range([props.height - props.margin.bottom, props.margin.top]);
+
+  const xAxis = d3.axisBottom(x).ticks(12);
+  const yAxis = d3.axisLeft(y).ticks((12 * props.height) / props.width);
```
Can you use your linear-scales to compute the tick positions instead (like the time example [here](https://github.com/d3/d3-axis/blob/master/README.md#axis_ticks))?

jakereps commented:
Thanks @ebolyen and @gregcaporaso! I'll get at these today!

jakereps commented:
Thanks for the review, @gregcaporaso! I've finished tweaking the Python code. I've left 2 variants of subsampling in place, with one method commented out. Is there a way to have these tested on a real dataset? The one left uncommented right now shows a minor speed increase over the other using the Atacama dataset, curious to see how it scales. The one thing not implemented is the bounds checking, but I had a question on how to address that. Should it fail hard when making the viz, or make the viz and display a warning on the plot page?

jakereps commented:
Thanks @ebolyen! Got most of the points addressed, just a few questions.

gregcaporaso commented:
@jakereps, I can pull down and test once you have all of your changes in for everyone else - does that work, or would it help to have me re-run this on real data sooner?

jakereps commented:
Sorry about all the push spam guys, but I think this is ready for a final pass and testing!

gregcaporaso commented:
I tested the latest version of this on real [single-end](https://dl.dropboxusercontent.com/u/2868868/temp/q2-demux-issue-38/demux.qzv) and [paired-end](https://dl.dropboxusercontent.com/u/2868868/temp/q2-demux-issue-38/demux-paired.qzv) data, as well as [PHRED64 data](https://dl.dropboxusercontent.com/u/2868868/temp/q2-demux-issue-38/phred64-demux.qzv) from the 2017.2 moving pictures tutorial. These look good to me!

gregcaporaso commented:
> Is there a way to have these tested on a real dataset?

@jakereps, it looks like the version that was commented (i.e., that uses `enumerate`) is a bit faster on real data. That took 3 minutes, 28 seconds while the version that uses `NamedTemporaryFile` took 4 minutes, 24 seconds.

I ran this twice to confirm the results, and they were similar each time.

jakereps commented:
Awesome, thanks @gregcaporaso! Swapped out to use the alternative subsampling method, changed the alerts, and added a "this plot was generated using x sequences" to the bottom of the page.

gregcaporaso commented:
@jakereps, there seems to be an issue with nan values in some plots, for example see boxes 1-5 [here](https://dl.dropboxusercontent.com/u/2868868/temp/q2-demux-issue-38/demux.qzv). This was generated from [this qza file](https://dl.dropboxusercontent.com/u/2868868/temp/q2-demux-issue-38/demux.qza). I'll keep looking into this tomorrow. It seems like the quality scores are being decoded ok, but we're getting nans back from pandas inside of `_compute_stats_of_df`.

gregcaporaso commented:
:beers:

Thanks everyone, especially @jakereps and @thermokarst, for all of the work on this awesome viz. The never-ending PR is merged!
